graylog2-server/src/main/java/org/graylog2/shared/plugins/PluginLoader.java

```diff
@@ -82,25 +85,27 @@ public PluginLoader(File pluginDir, ChainingClassLoader classLoader) {
         }
 
         LOG.debug("Loading [{}] plugins", files.length);
-        final ImmutableSet.Builder<Plugin> plugins = ImmutableSet.builder();
-        for (File jar : files) {
-            if (!jar.isFile()) {
-                LOG.debug("{} is not a file, skipping.", jar);
-            } else {
-                try {
-                    LOG.debug("Loading <" + jar.getAbsolutePath() + ">");
-                    final ClassLoader pluginClassLoader = URLClassLoader.newInstance(new URL[]{jar.toURI().toURL()});
-                    classLoader.addClassLoader(pluginClassLoader);
-                    final ServiceLoader<Plugin> pluginServiceLoader = ServiceLoader.load(Plugin.class, classLoader);
-
-                    plugins.addAll(pluginServiceLoader);
-                } catch (MalformedURLException e) {
-                    LOG.error("Cannot open JAR file for discovering plugins", e);
-                }
-            }
-        }
-
-        return plugins.build();
+        final List<URL> urls = Lists.newArrayList(files).stream()
+                .filter(File::isFile)
+                .map(jar -> {
+                    try {
+                        LOG.debug("Loading <" + jar.getAbsolutePath() + ">");
+                        return jar.toURI().toURL();
+                    } catch (MalformedURLException e) {
+                        LOG.error("Cannot open JAR file for discovering plugins", e);
+                        return null;
+                    }
+                })
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        // Create one URL class loader for all plugins so they can see each other.
+        // This makes plugin inter-dependencies work.
+        classLoader.addClassLoader(URLClassLoader.newInstance(urls.toArray(new URL[urls.size()])));
+
+        final ServiceLoader<Plugin> pluginServiceLoader = ServiceLoader.load(Plugin.class, classLoader);
+
+        return ImmutableSet.<Plugin>builder().addAll(pluginServiceLoader).build();
```

Comment:
This could simply be written as `return ImmutableSet.copyOf(pluginServiceLoader);`.